mypy/messages.py

```diff
@@ -860,6 +869,20 @@ def argument_incompatible_with_supertype(
         self.fail('Argument {} of "{}" incompatible with {}'
                   .format(arg_num, name, target), context)
 
+        if name == "__eq__":
+            assert isinstance(context, FuncDef)
+            multiline_msg = self.comparison_method_example_msg(class_name=context.info.name())
```

Comment:
Getting anything except line/column from `context` is a bad idea. First, this kind of defeats the purpose of the type annotation. Second, what if this is an `OverloadedFuncDef`, or a `Decorator`?

I think it is better to add an optional argument `current_type_name` to `argument_incompatible_with_supertype`, and `if name == '__eq__': assert current_type_name is not None` etc. (Or maybe even make it a required argument if there are only couple of call sites to update).